### PR TITLE
Fixed type predicate inference for discriminated union parameters

### PR DESCRIPTION
**Pull request by Andarist**
fixes https://github.com/microsoft/TypeScript/issues/57947

### CODE REVIEW

**Comment by RyanCavanaugh** [comment]
@typescript-bot test it

**Comment by RyanCavanaugh** [comment]
```ts
              const suites = tasksGroup.filter(group => group.type === 'suite')
              const tests = tasksGroup.filter(group => group.type === 'test')
              const groups = shuffle([suites, tests], sequence.seed)
              tasksGroup = groups.flatMap(group => shuffle(group, sequence.seed))
```
Seems like an expected break

**Comment by Andarist** [comment]
Full TS playground for the above: [TS playground](https://www.typescriptlang.org/play?target=99&ts=5.5.0-dev.20240327#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwGcALZRRCEAHgBUA+AClhigE8AueagbQF0AaQiBDAA-B1TIAtgCMQMAJQduPANwBYAFCasqDHMRQwCAMrIse+AG9N8eBhYAHEBwBEBM3pfqNAX03bdfUMEahACDCsbO0dneBc9cK9NPy0NHT0YAyN4AGFkcJxJSI1beydXMHyMQqTffw0ykKgCAGt4AF5OMIiAH3hTcwQ+vILJb01QSFgECgiCEABHZBB0EGMhYHEpWRhvSeg4eFm7ZpaCAHEYHGQHJVPecY0Aeif4Spg4XXh6HUQ5OGA8GkOBwFCg+GBoJA4PgcAwyBg+EaBHgOjsRBw8yQWAgGQI8k0KHQ2Dwdm6AEZ6PJirYwHhwoQPGEOidWhcrjcAHSIHEZej0ADmHIc1PatHgQuuDk5jQ67U6biZLnk3lp9IiCQwKM6GFO7Kl3N5cn5kpuovFpulsvlCs1ytVb3VEuF2sIJDIFHoXHcgwIAk1BH4giWKyM62EKqiurZlylLMtBG50AwAFkoA4TcLzW7SOQQILhQJ5iHVuHgPJI3VUi94M13JJhKjJA4rgA3Ru-OShhD0Hm4uQo5rRJzwFvCLBgKAJAkaImYXBI7oAJipNMdqAZPoSLOjZ1jXL7fKitgLUsUzrjWBRAwsYovXOt8riW5ALj4UUrao3Gu6rt3+oPI0YHoY9vktc9LVRFFQgZO9LRlGI5Vtbo3w-B06W-e8HFdYhc09b0mT9MlwkDItFmWUsNk-Vk92FeMXSTKc0wzU8zQ6cVcI9fNLTIkswyoysUk0GsYWhAgWHgHAHAia4ZMQdEEFQEAAHcgRAIgoFbXAYCBCTwVQHBdWwVABQU2EQHhRFh2ZIdITBDdUSRDEsUPORRwBCcpzCQk0HnUlNQAZlXawSnXTdCJ3PV92lVzgNYkUODs6F8Dg4UEJHG1nyVaiMIZANIpjNLYszM9EpBez2Kw9KEEy+IUJyp0ExZTi8y9F8iIDINiwo-iIwdf9ovoqVE3IJj0xKti7xaz0eODHq1gE7whOeV4CCwSQcVgCAJLwBBJ3wZScBgFpYGuVBAQwIgr10xkHAcbadFMqBrNHWAoAbDI7BweAAANppAH6fOJBdiIwAAWYKolyuYIp1KKiqAkDQpPcCOEg66bxqi00sfBUX1Q0KGsw-K4cKg1itA+KILo66YIiVKDVxuI7XfQn0Mal1mvdVqCN9f1fy68juzLaiBrozoE0Y1Nxqpyr-poU4GFm7rhcW5J6hAAAPBwjoiSwfG8IA)

**Review by RyanCavanaugh — APPROVED** [review]

**Comment by RyanCavanaugh** [comment]
@danvk look right to you?

**Review by danvk — COMMENTED** [review]
LGTM

I'd wondered what the right value for `declaredType` in the false check was but didn't have a test case that showed it making a difference. Now we do! Thanks for tracking this down, @Andarist.

Not relevant to merging this PR, but I am curious why `declaredType` matters in this case but not most others. Presumably there are other, unreported bugs that this fixed.

**Comment by Andarist** [comment]
> Not relevant to merging this PR, but I am curious why declaredType matters in this case but not most others. Presumably there are other, unreported bugs that this fixed.

Without this TS won't attempt to discriminate the union since unions are not involved, see the condition [here](https://github.dev/microsoft/TypeScript/blob/824cd6eb314bdcb8b7bde158025c82ebc014db0a/src/compiler/checker.ts#L28416). 

The fix relies on the fact that the assigned type might be different than the declared type and it's still "preserved":
```ts
type U = { type: "foo" } | { type: "bar" };

const item: U = { type: "foo" };

item;
// ^? const item: { type: "foo"; }
```

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -37683,8 +37683,8 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             { flags: FlowFlags.Start };
         const trueCondition: FlowCondition = {
             flags: FlowFlags.TrueCondition,
-            node: expr,
             antecedent,
+            node: expr,
```
[the remark this answers is not in the document]
Comment (Andarist) [reply src/compiler/checker.ts:37687]: I tried to follow the structure used by the ones created in the binder to keep functions receiving this monomorphic. I'm not sure if my change prompted [this further experiment](https://github.com/microsoft/TypeScript/pull/57977) but it certainly looked like it to me since it was open soon after I opened this one here ;p

TLDR: this was meant as an optimization for the engine. I didn't measure it at all since I don't have time for it but it's usually better to create structures in the same way (order of keys etc matters for the engine)

[the remark this answers is not in the document]
Comment (danvk) [reply src/compiler/checker.ts:37687]: 👍 Makes sense. The calls to `getFlowTypeOfReference` are so expensive compared to everything else in `getTypePredicateFromBody` that I doubt micro-optimizations matter (see https://github.com/microsoft/TypeScript/pull/57465#issuecomment-1974921974 and https://github.com/microsoft/TypeScript/pull/57660). But I can't imagine they hurt!

[the remark this answers is not in the document]
Comment (Andarist) [reply src/compiler/checker.ts:37687]: once u apply micro-optimizations at scale there are some good wins to be had :P https://github.com/microsoft/TypeScript/pull/57977#issuecomment-2032268884
